source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,183 @@
+
+void GzipFilter::onDestroy() {}
+
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  // TODO(gsagula): The current implementation checks for the presence of 'gzip' and if the same
+  // is followed by Qvalue. Since gzip is the only available encoding right now, order/priority of
+  // preferred server encodings is disregarded (RFC2616-14.3).
+  skip_compression_ =
+      !(headers.AcceptEncoding() &&
+        std::regex_search(headers.AcceptEncoding()->value().c_str(), acceptEncodingRegex()));
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  response_headers_ = &headers;
+  if (!end_stream && !skip_compression_ && isMinimumContentLength() && isContentTypeAllowed() &&
+      isCacheControlAllowed() && isEtagAllowed() && isLastModifiedAllowed() &&
+      isTransferEncodingAllowed() && !headers.ContentEncoding()) {
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isMinimumContentLength() const {
+  if (response_headers_->ContentLength()) {
+    uint64_t content_length{};
+    StringUtil::atoul(response_headers_->ContentLength()->value().c_str(), content_length);
+    return content_length >= config_->minimumLength();
+  }
+
+  return response_headers_->TransferEncoding() &&
+         response_headers_->TransferEncoding()->value().find(
+             Http::Headers::get().TransferEncodingValues.Chunked.c_str());
+}
+
+bool GzipFilter::isContentTypeAllowed() const {
+  if (config_->contentTypeValues().size() && response_headers_->ContentType()) {
+    return std::any_of(config_->contentTypeValues().begin(), config_->contentTypeValues().end(),
+                       [this](const auto& value) {
+                         return response_headers_->ContentType()->value().find(value.c_str());
+                       });
+  }
+  return true;
+}
+
+bool GzipFilter::isCacheControlAllowed() const {
+  if (config_->cacheControlValues().size() && response_headers_->CacheControl()) {
```

Comment:
why would cache-control influence the compressibility of something?  Except maybe cache-control:no-transform.

Reply:
It does not affect compression, but the necessity for such. Since the response is going to be cached, there are situations where skip compression might be desirable. Less common I believe, envoy could be behind another proxy that might cache certain content and then serve it compressed.

Reply:
RE "response is going to be cached", I think that's what Vary:Accept-Encoding is for.  HTTP caches should be able to hold a gzipped response and serve it to a non-gzip-accepting client by decompressing it on the fly, if that header is in the cached response.  https://blog.stackpath.com/accept-encoding-vary-important is worth a read.

RE "behind another proxy" -- I think in that case the site admin would just disable gzipping in Envoy then, right?  Why would it need to be based on cache-control headers?

I'm not objecting to this setting, I just want to better understand what the usage model is and its relationship to caches.

Reply:
👀

Reply:
Thanks for pointing this out Joshua. 
I think that Vary should be also available in the configuration. In regards to cache-control, I was thinking more in in terms of proxied requests, like Nginx does on the presence of Via request header. See gzip_proxied at http://nginx.org/en/docs/http/ngx_http_gzip_module.html
Thoughts?

Reply:
OK, sounds like you are going for a drop-in replacement for nginx gzip support? Fair enough.  I don't know why someone would configure cache-control's effect on gzipping in an nginx proxy either, but it must have been put in for a reason.  In that case, let's just do it efficiently.   In the case of cache-control, I think it means splitting any number of attributes on ",", and doing a hash lookup in your set on each of them.  E.g. `cache-control: no-cache, no-store, must-revalidate, max-age=0` should (IMO) result in 4 exact hash lookups.

The most relevant setting (AFAICT), cache-control:no-transform, is not mentioned in the nginx doc you referenced, and I don't see it in your PR either.  By default, IMO, you should honor cache-control:no-transform and avoid gzipping any response with that header.

I think it's very important for proxies to get the defaults right, especially Vary:accept-encoding, as messing these up can result in hard-to-diagnose downstream failures.

Reply:
Good point on having ```cache-control:no-transform``` by default. I tried to make the feature as least opinionated as possible, so the users could opt for skipping compression on whatever cache-control value/s they need to, including no-transform.
I will work on these changes, thanks!

Reply:
I'd defer to @htuch and @mattklein123 and @alyssawilk on the design philosophy here, as I am still new to Envoy development.  A few options are:

1. If nginx's gzip module does it, then Envoy's should too.
2. Provide mechanism not policy about how the proxy should behave, and let the config figure out what to do (ie "least opinionated as possible)
3. Provide a paved path toward a few different sensible scenarios

There is some subtlety here that I think is worth getting right in the C++ code, so I kind of favor option 3, and these scenarios come to mind:

a. Enable gzip with some level of compression aggressiveness, respecting cache-control:no-transform from origin, and adding vary:accept-encoding to the response to the client.
b. Disable gzip completely, e.g. because the Envoy being configured is sitting in a high-bandwidth rack and some other proxy will gzip en route to client.

But I totally understand if the market for Envoy wants it to be a clean drop-in replacement for nginx and whatever features that has in its gzip setup.

The challenge with #2 above (least opinionated) is that you then need the person configuring Envoy to know the HTTP semantics and best practices, and enforce them via config.